File: python_programs/thread_context_switch.py
import threading
import time

NUM_ITERATIONS = 1000

def worker():
    for _ in range(NUM_ITERATIONS):
        pass

total_time = 0

for _ in range(10):
    thread1 = threading.Thread(target=worker)
    thread2 = threading.Thread(target=worker)
    start = time.perf_counter_ns()
    thread1.start()
    thread2.start()
    thread1.join()
    thread2.join()
    end = time.perf_counter_ns()
    total_time += (end - start)

with open("./results/python_thread_context_switch_results.txt", "w") as f:
    f.write(f"Average: {total_time / 10:.6f} ns\n")

File: java_programs/ThreadContextSwitch.java
import java.io.FileWriter;
import java.io.IOException;

public class ThreadContextSwitch {
    private static final int NUM_ITERATIONS = 1000;

    public static void main(String[] args) throws IOException {
        long totalTime = 0;
        for (int test = 0; test < 10; test++) 
        {
            Thread worker1 = new Thread(() -> 
            {
                for (int i = 0; i < NUM_ITERATIONS; i++);
            });
            Thread worker2 = new Thread(() -> 
            {
                for (int i = 0; i < NUM_ITERATIONS; i++);
            });
            long start = System.nanoTime();
            worker1.start();
            worker2.start();
            try 
            {
                worker1.join();
                worker2.join();
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
            long end = System.nanoTime();
            totalTime += (end - start);
        }

        try (FileWriter writer = new FileWriter("./results/java_thread_context_switch_results.txt")) {
            writer.write(String.format("Average: %.6f ns\n", totalTime / 10.0));
        }
    }
}
